[WOC] Implement Unfinished Business

Codermann63 opened the pull request:
Tested by hand.
I will make unit tests if asked.

xenohedron commented on Mage.Sets/src/mage/cards/u/UnfinishedBusiness.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    @Override
+    public Effect copy() {
```
Return type of the method should be `UnfinishedBusinessEffect` rather than just `Effect`.

xenohedron commented on Mage.Sets/src/mage/cards/u/UnfinishedBusiness.java:
```diff
@@ -0,0 +1,123 @@
+
+    public UnfinishedBusinessEffect() {
+        super(Outcome.PutCreatureInPlay);
+        staticText = "Return target creature card from your graveyard to the battlefield, then return up to two target Aura and/or Equipment cards from your graveyard to the battlefield attached to that creature. (If the Auras can not enchant that creature, they remain in your graveyard.)";
```
use `<i>` `</i>` around the reminder text

xenohedron commented on Mage.Sets/src/mage/cards/u/UnfinishedBusiness.java:
```diff
@@ -0,0 +1,123 @@
+
+class UnfinishedBusinessEffect extends OneShotEffect{
+
+    public UnfinishedBusinessEffect() {
```
card-specific classes like this should be package-private (i.e. remove the `public` access modifier)

xenohedron commented on Mage.Sets/src/mage/cards/u/UnfinishedBusiness.java:
```diff
@@ -0,0 +1,123 @@
+        if (targetCreature != null){
+            controller.moveCards(targetCreature, Zone.BATTLEFIELD, source, game);
+        }
+        Permanent permanentCreature = targetCreature == null ? null : game.getPermanent(source.getTargets().getFirstTarget());
```
Not really sure why you need this ternary; I'd expect there to be no issue with getPermanent returning null. I'm also mildly surprised that it can find the permanent from the target after it changed zones. Usually I'd see something like `Permanent permanentCreature = game.getPermanent(targetCreature.getId())` but if what you wrote works then I guess it's fine.

Codermann63 replied:
Correct me if i'm wrong, if I use targetCreature.getId() and targetCreature is null, i would get a null pointer exception? So i would need the ternary in that case?
Reminder: I get targetCreature from game.getCard(...)

xenohedron replied:
That's right, you have to nullcheck when calling a method on it. Sorry I didn't think that rambly feedback all the way through, but looks good now.

xenohedron reviewed (commented):
Only a couple minor nits.

Can you please write two test cases?
1. Return a creature with shroud, and return an Aura and an Equipment checking that both attach. (My interpretation of the rules is that the attaching doesn't target the permanent.)
2. Return a creature with protection from a color, and try to return an aura and an equipment of that color. The aura should remain in graveyard and the equipment should return but not attach. Suggest White Knight, Ghoulflesh, Enormous Energy Blade (to check that the creature isn't tapped).

(let me know if you think I'm misinterpreting the rules and something different should happen, or if the engine doesn't support the correct behavior)

Codermann63 commented:
I added the unit tests [xenohedron](https://github.com/xenohedron) suggested and made the small changes suggested.
I also added one to check:
If the targeted creature gets exiled before the resolution of Unfinished Business, then equipment is still returned.

I have a question in line 82 and 102 of UnfinishedBusiness.java i return false. Is this correct?

xenohedron commented:
Re: lines 82 and 102- in general the return value of apply method should be true if something happened and false if nothing happened. However, nothing depends on the return value of this particular apply method (it has no subclasses trying to use its logic). So it's fine as is.

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/serverside/base/impl/CardTestPlayerAPIImpl.java:
```diff
@@ -1572,6 +1572,23 @@ public void assertIsAttachedTo(TestPlayer thePlayer, String theAttachment, Strin
 
+    }
+
+    public void assertIsNotAttachedTo(TestPlayer thePlayer, String theAttachment, String thePermanent) {
```
This can be a separate scope, but probably we should refactor to a single method `assertAttachedTo` that takes a boolean parameter

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/cards/single/woc/UnfinishedBusinessTest.java:
```diff
@@ -0,0 +1,108 @@
+        castSpell(1,PhaseStep.PRECOMBAT_MAIN,playerA,UNFINISHEDBUSINESS,SHROUDCREATURE);
+        addTarget(playerA, AURA+"^"+EQUIPMENT);
+        waitStackResolved(1, PhaseStep.PRECOMBAT_MAIN);
+
```
use `setStrictChooseMode(true)` for all new tests

xenohedron commented on Mage.Sets/src/mage/cards/u/UnfinishedBusiness.java:
```diff
@@ -0,0 +1,125 @@
+
+        if (targetCreature != null){
+            controller.moveCards(targetCreature, Zone.BATTLEFIELD, source, game);
+        }
```
We might want a `game.getState().processAction(game)` in here, to guard against edge cases where continuous effects determine whether the Aura can be attached to the creature.

Very contrived example: return Grim Strider while you have four cards in hand (making it a 2/2), Runner's Bane should attach to it.

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/cards/single/woc/UnfinishedBusinessTest.java:
```diff
@@ -0,0 +1,108 @@
+
+public class UnfinishedBusinessTest extends CardTestPlayerBase {
+
+    private static final String UNFINISHEDBUSINESS = "Unfinished Business";
```
helps to add comments with the text of the cards used

xenohedron reviewed (commented):
looking good!

xenohedron reviewed (approved):
Great work. Thanks for the contribution.
